clang/test/CodeGenCUDA/device-use-host-var.cu

```diff
@@ -104,3 +106,17 @@ void fun() {
   (void) b<double>;
   (void) var_host_only;
 }
+
+// NEG-NOT: external_func
+extern __global__ void external_func();
+// NEG-NOT: @external_dep
+extern void* const external_dep[] = {
+  (void*)(external_func)
+};
+// NEG-NOT: @external_arr
```

Comment:
Not sure this negative check really works as expected; won't any global variable be printed at the top of the file before any functions?

Reply:
Not sure I got it correctly, but AFAIU in this test we are checking only the device code.
So with the current patch an external global variable should not be included to the device code (only to the host code) if it's not used in any device functions at the current compilation unit. If it's actually used 
Though if this variable is used at the device function than it should be manually marked as __constant__ variable. I can add this test case as well if it's needed.

Reply:
I mean negative checks are inherently fragile, and the place you are checking it does not exist isn't where globals are printed. An -implicit-check-not=xxx argument to FIleCheck would be more thorough

Reply:
Ah, sorry misunderstood original comment.
Changed test code to use implicit-check, thanks for the reference